### ISSUE TITLE
throwing non-errors (specifically strings)

### ISSUE THREAD

**Issue by jonathanong** [eslint/eslint#1791]
is there a rule to make sure people don't throw strings?

**Comment by nzakas** [comment]
We do not. For future reference, our rules are all documented: http://eslint.org/docs/rules/

**Comment by jonathanong** [comment]
Well this is a feature request then :)

**Comment by nzakas** [comment]
Please take a moment to read over what new rule proposals should include: http://eslint.org/docs/developer-guide/contributing.html#new-rules

**Comment by doberkofler** [comment]
I would very much like to see this as a build-in rule.

**Comment by nzakas** [comment]
Please put together a proposal for the rule per the link.

**Comment by doberkofler** [comment]
I must start by saying that I have very strict guidelines on what can be thrown, so this might be far to restrictive for others.

I would want this rule to basically enforce only throwing errors in the form `throw new Error()` where Error should be one of an arbitrary list of constructors that can be specified in the rule definition.

The rule should be disabled by default, could be part of the `Best Practices` and could be specified with something like this:
```
"throw-style": [1, { "force-new": true, "constructors": ["Error", "TypeError", "CustomError"]}]
```

**Comment by gcochard** [comment]
:+1:

**Comment by nzakas** [comment]
There's a lot more details to hash out here. We can easily flag patterns like:

```js
throw "foo";
```

We can also notice

```js
throw new Error();
```

What we can't easily do is notice

```js
var err = new Error();
throw err;
```

I don't know that it's worth trying to limit the constructor being used given that.

**Comment by doberkofler** [comment]
As I mentioned, I would be fine to only allow `throw new Error()` as I have very strict rules as I almost never use any other form of exception. Clearly for others this might be to restrictive.
Looking at the code I see around throwing a variable does not seem to very common.

**Comment by nzakas** [comment]
The larger question around this rule is this: should it disallow specific patterns or allow specific patterns?

**Comment by doberkofler** [comment]
I would only allow one pattern

**Comment by ilyavolodin** [comment]
If it will allow only one pattern (throw new Error), you will have issues with code like this:
```js
try {
   ....
} catch (e) {
   ....
   throw e;
}
This feels specific enough to your coding style to be a custom rule, instead of going into main repo.

**Comment by nzakas** [comment]
Only allowing one pattern is far too limiting. I'm leaning more towards looking for specific bad examples and warning them.

**Comment by doberkofler** [comment]
I agree with @ilyavolodin that the pattern
```javascript
try {
   ....
} catch (e) {
   ....
   throw e;
}
```
is so common that it should also be allowed, but I'm not so sure if there are other reasonably common code pattern for throwing exceptions.

Am I guessing right, that supporting the above pattern would be be possible in eslint by matching the argument `e` with the object `e` that will be thrown ?

**Comment by nzakas** [comment]
I'm not in favor of ESLint detecting such a pattern. We know from experience that there are frequently an unknown number of exception cases when we try to detect things, which is why we opt for detecting small, known cases rather than trying to account for everything.

I'm getting closer to being convinced that we should simply detect `throw` followed by a literal and then warn that an error should be used.

**Comment by xjamundx** [comment]
Oh ha I already built this a few weeks ago. I should have mentioned it :)
